node/pkg/driver/node.go

```diff
@@ -345,47 +341,107 @@ func (d *NodeService) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+	case *csi.VolumeCapability_Block:
+		isFSVolume = false
+	}
+	isTargetPathExists := d.NodeUtils.IsPathExists(targetPathWithHostPrefix)
+	if isTargetPathExists {
+		// check if already mounted
+		isMounted, err := d.isTargetMounted(targetPath, isFSVolume)
+		if err != nil {
+			logger.Debugf("Existing mount check failed {%v}", err.Error())
+			return nil, err
+		}
+		if isMounted { // idempotent case
+			return &csi.NodePublishVolumeResponse{}, nil
 		}
 	}
+	if isFSVolume {
+		fsType := volumeCap.GetMount().FsType
+		err = d.mountFileSystemVolume(mpathDevice, targetPath, fsType, isTargetPathExists)
+	} else {
+		err = d.mountRawBlockVolume(mpathDevice, targetPath, isTargetPathExists)
+	}
 
-	// if the device is not mounted then we are mounting it.
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	logger.Debugf("NodePublishVolume Finished: multipath device is now mounted to targetPath.")
 
-	volumeCap := req.GetVolumeCapability()
-	fsType := volumeCap.GetMount().FsType
+	return &csi.NodePublishVolumeResponse{}, nil
+}
 
+func (d *NodeService) mountFileSystemVolume(mpathDevice string, targetPath string, fsType string, isTargetPathExists bool) error {
 	if fsType == "" {
 		fsType = defaultFSType
 	}
-
-	if _, err := os.Stat(targetPathWithHostPrefix); os.IsNotExist(err) {
-		logger.Debugf("Target path directory does not exist. creating : {%v}", targetPathWithHostPrefix)
-		d.mounter.MakeDir(targetPathWithHostPrefix)
+	logger.Debugf("Volume will be have FS type : {%v}", fsType)
+	targetPathWithHostPrefix := GetPodPath(targetPath)
+	if !isTargetPathExists {
+		logger.Debugf("Target path directory does not exist. Creating : {%v}", targetPathWithHostPrefix)
+		err := d.Mounter.MakeDir(targetPathWithHostPrefix)
+		if err != nil {
+			return status.Errorf(codes.Internal, "Could not create file %q: %v", targetPathWithHostPrefix, err)
+		}
 	}
-
 	logger.Debugf("Mount the device with fs_type = {%v} (Create filesystem if needed)", fsType)
+	return d.Mounter.FormatAndMount(mpathDevice, targetPath, fsType, nil) // Passing without /host because k8s mounter uses mount\mkfs\fsck
+}
 
-	logger.Debugf("FormatAndMount start [goid=%d]", util.GetGoID())
-	err = d.mounter.FormatAndMount(mpathDevice, targetPath, fsType, nil) // Passing without /host because k8s mounter uses mount\mkfs\fsck
-	// TODO: pass mount options
-	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
+func (d *NodeService) mountRawBlockVolume(mpathDevice string, targetPath string, isTargetPathExists bool) error {
+	logger.Debugf("Raw block volume will be created")
+	targetPathWithHostPrefix := GetPodPath(targetPath)
+	// Create mount file and its parent directory if they don't exist
+	targetPathParentDirWithHostPrefix := filepath.Dir(targetPathWithHostPrefix)
+	if !d.NodeUtils.IsPathExists(targetPathParentDirWithHostPrefix) {
+		logger.Debugf("Target path parent directory does not exist. creating : {%v}", targetPathWithHostPrefix)
```

Comment:
https://github.com/IBM/ibm-block-csi-driver/pull/117#discussion_r373045255
this is one of those other places, please re-review your messages values!

Reply:
already fixed in other places too

Reply:
why is this resolved? I don't see any change here

Reply:
Please b e m ore specific.
 I thought you mean targetPathWithHostPrefix instead of targetPath

Reply:
Oh sorry, I missed that diff. Why would I mean `targetPathWithHostPrefix` if you are not using it for `IsPathExists`?

Reply:
I still don't think this is resolved. Please continue the conversation if you think otherwise.